core-java-modules/core-java-concurrency-advanced-4/src/main/java/com/baeldung/producerconsumer/Producer.java

```diff
@@ -18,20 +18,24 @@ public void run() {
 
     public void produce() {
         while (runFlag) {
-            Message message = generateMessage();
-            while (dataQueue.isFull()) {
-                try {
-                    dataQueue.waitOnFull();
-                } catch (InterruptedException e) {
-                    e.printStackTrace();
+            synchronized (this) {
+                while (dataQueue.isFull()) {
+                    try {
+                        dataQueue.waitOnFull();
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                        break;
+                    }
+                }
+                if (!runFlag) {
                     break;
                 }
             }
-            if (!runFlag) {
-                break;
+            Message message = generateMessage();
```

Comment:
Move 34-37 inside the above synchronized block, like in Consumer

Reply:
Done